### PR TITLE
[idea] Indentation adjustment based on the block model

### PR DESCRIPTION
**Pull request by akosyakov**
Change-Id: Ia162097ef016db58e9204060b7206ddf4b9fd0b7
Signed-off-by: akosyakov <anton.kosyakov@itemis.de>

### CODE REVIEW

**Comment by akosyakov** [comment]
Jenkins intellij-tests-master-latest

File: intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/formatting/DefaultCommenter.xtend [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		extension ITokenDefProvider tokenDefProvider
+	) {
+		val mlCommentEntry = tokenDefMap.entrySet.findFirst[value == tokenName]
+		if (mlCommentEntry != null)
+			mlCommentEntry.key.getIElementType
+	}
+
+	// TODO: extract to a configurable constant
+	override getLineCommentPrefix() {
+		'//'
```
Comment (szarnekow) [inline intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/formatting/DefaultCommenter.xtend]: The is the org.eclipse.xtext.ui.editor.toggleComments.SLCommentPrefixCalculator - mybe it helps here?

File: intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/formatting/DefaultXtextBlock.xtend [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+import static extension com.intellij.psi.formatter.FormatterUtil.*
+import static extension com.intellij.psi.impl.source.tree.TreeUtil.*
+
+/**
+ * @author kosyakov - Initial contribution and API
+ */
+class DefaultXtextBlock extends AbstractBlock implements ModifiableBlock {
+
+	static val char LINE_FEED = '\n'
```
Comment (szarnekow) [inline intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/formatting/DefaultXtextBlock.xtend]: That looks as if it would break on windows?

Comment (akosyakov) [inline intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/formatting/DefaultXtextBlock.xtend]: It seems that IDEA abstracts away from platform specific line separators and inserts them only on save, as a client we should always use only '\n'.

Look at 'What are the rules of working with it?' here: https://confluence.jetbrains.com/display/IDEADEV/IntelliJ+IDEA+Architectural+Overview

File: intellij/org.eclipse.xtend.idea.tests/src/org/eclipse/xtend/idea/autoedit/AbstractCStyleLanguageAutoEditTest.xtend [hunk 3]
```diff
@@ -555,63 +567,63 @@ abstract class AbstractCStyleLanguageAutoEditTest extends LightToolingTest {
...
-		assertState("  /* foo \n   * |\n   */")
+		assertState("  /* foo \n   * | */")
 	}
 
 	def void testMLComments_09() {
 		configureByText("/* foo |*/")
 
 		myFixture.type('\n')
-		assertState("/* foo \n * |*/")
+		assertState("/* foo \n* |*/")
```
Comment (svenefftinge) [inline intellij/org.eclipse.xtend.idea.tests/src/org/eclipse/xtend/idea/autoedit/AbstractCStyleLanguageAutoEditTest.xtend]: I think it is a bug in idea, that it doesn't always add the whitespace before the '*'.

**Pull request merged**